### PR TITLE
feat: `toHaveProp` matcher

### PR DESCRIPTION
**Pull request by AntoineThibi**
### Summary

Implement `toHaveProp` matcher

### Test plan

`yarn test`

### CODE REVIEW

File: src/matchers/__tests__/to-have-prop.test.tsx
Comment (pierrezimmermannbam) [inline src/matchers/__tests__/to-have-prop.test.tsx:34]: I guess it was like that in jest-native matcher but wouldn't it be better to use not rather than expect the contrary to throw?

Comment (pierrezimmermannbam) [reply src/matchers/__tests__/to-have-prop.test.tsx:34]: Oh I realize now using this could allow to test the message, can you replace to throw by toThrowErrorMatchingInlineSnapshot()?

File: src/matchers/to-have-prop.ts
Comment (pierrezimmermannbam) [inline src/matchers/to-have-prop.ts:21]: You should use the already existing method `checkHostElement` and delete this one

File: src/matchers/utils.tsx
Comment (pierrezimmermannbam) [inline src/matchers/utils.tsx:125]: I think all of the code added here can be removed if you use checkHostElement instead

**Review by pierrezimmermannbam — CHANGES_REQUESTED** [review]
Hello @AntoineThibi, thanks for the contribution! I've left some comments to address before this can be merged

File: src/matchers/__tests__/to-have-prop.test.tsx
Comment (mdjastrzebski) [reply src/matchers/__tests__/to-have-prop.test.tsx:4]: Imo that could be moved to jest-setup.js assuming that it will work ;-)

File: src/matchers/to-have-prop.ts
Comment (mdjastrzebski) [reply src/matchers/to-have-prop.ts:21]: Yes, `checkHostElement` is the check to use

File: src/matchers/utils.tsx
Comment (mdjastrzebski) [inline src/matchers/utils.tsx:129]: ditto: pls use `formatMessage` already present in this file instead.

File: src/matchers/utils.tsx
Comment (mdjastrzebski) [inline src/matchers/utils.tsx:143]: ditto: pls use `checkHostComponent` already present in this file instead.

File: src/matchers/utils.tsx
Comment (mdjastrzebski) [inline src/matchers/utils.tsx:176]: ditto: pls use `checkHostComponent` already present in this file instead.

File: src/matchers/to-have-prop.ts
Comment (mdjastrzebski) [inline src/matchers/to-have-prop.ts:12]:
```suggestion
  return value === undefined
    ? `Element should have prop ${name}`
    : `Element should have prop ${name} with value ${stringify(value)}`;
```

File: src/matchers/to-have-prop.ts
Comment (mdjastrzebski) [inline src/matchers/to-have-prop.ts:23]:
```suggestion
  const propValue = element.props[name];
```

File: src/matchers/__tests__/to-have-prop.test.tsx
Comment (mdjastrzebski) [inline src/matchers/__tests__/to-have-prop.test.tsx:16]: [nit] queryBy => getBy

File: src/matchers/__tests__/to-have-prop.test.tsx
Comment (AntoineThibi) [reply src/matchers/__tests__/to-have-prop.test.tsx:4]: It doesn't work with the `extend-expect.test` which checks that those new methods are not present in the expect by default.

**Comment by AntoineThibi** [comment]
@pierrezimmermannbam @mdjastrzebski I've made the modifications !

File: src/matchers/__tests__/to-have-prop.test.tsx
Comment (pierrezimmermannbam) [reply src/matchers/__tests__/to-have-prop.test.tsx:4]: Oh indeed I hadn't thought about that. I can't think of a simple solution to solve this, creating a jest config only for this file seems overkill for what we're trying to achieve. In that case either we remove that test or we keep the imports, but I think it's better to just keep the imports, this test is rather valuable and having to add imports is not a huge downside

**Review by pierrezimmermannbam — APPROVED** [review]
That's ok for me, thanks @AntoineThibi ! Just need to wait for @mdjastrzebski 's approval before we merge

**Review by mdjastrzebski — APPROVED** [review]
@AntoineThibi great work! 💯 I've done some final polishing before merging.

If you have spare time please pick some more Jest matchers from #1468.

**Comment by mdjastrzebski** [comment]
🎉 Release in version: [v12.4.0](https://github.com/callstack/react-native-testing-library/releases/tag/v12.4.0)

**Pull request merged**
